modules/effects/arrow-effects-data/src/main/kotlin/arrow/effects/internal/CancelablePromise.kt

```diff
@@ -0,0 +1,126 @@
+package arrow.effects.internal
+
+import arrow.Kind
+import arrow.core.*
+import arrow.effects.Promise
+import arrow.effects.typeclasses.Concurrent
+import arrow.effects.typeclasses.mapUnit
+import java.util.concurrent.atomic.AtomicReference
+
+internal class CancelablePromise<F, A>(CF: Concurrent<F>) : Promise<F, A>, Concurrent<F> by CF {
+
+  internal sealed class State<out A> {
+    data class Pending<A>(val joiners: Map<Token, (Either<Throwable, A>) -> Unit>) : State<A>()
+    data class Full<A>(val value: A) : State<A>()
+    data class Error<A>(val throwable: Throwable) : State<A>()
+  }
+
+  private val state: AtomicReference<State<A>> = AtomicReference(State.Pending(emptyMap()))
+
+  override fun get(): Kind<F, A> = defer {
+    when (val current = state.get()) {
+      is State.Full -> just(current.value)
+      is State.Pending -> cancelable { cb ->
+        val id = unsafeRegister(cb)
+        delay { unregister(id) }
+      }
+      is State.Error -> raiseError(current.throwable)
+    }
+  }
+
+  override fun tryGet(): Kind<F, Option<A>> = delay {
```

Comment:
It's a safe operation since it returns `Option`. Documentation is not provided in the internal implementations, it's explained in the `Promise` interface.